go/pkg/client/bytestream.go

```diff
@@ -26,33 +48,52 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
 	return err
 }
 
+// WriteBytesWithOptions uploads a byte slice with a given resource name to the CAS
+// at an arbitrary offset but retries still resend from the specified offset.
+func (c *Client) WriteBytesWithOptions(ctx context.Context, name string, data []byte, opts ...ByteStreamWriteOption) (int64, error) {
+	ue := uploadinfo.EntryFromBlob(data)
+	ch, err := chunker.New(ue, false, int(c.ChunkMaxSize))
+	if err != nil {
+		return 0, err
+	}
+	writtenBytes, err := c.writeChunked(ctx, name, ch, opts...)
+	if err != nil {
+		return 0, err
+	}
+	return writtenBytes, nil
+}
+
 // writeChunked uploads chunked data with a given resource name to the CAS.
-func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker) (int64, error) {
+func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker, rawOpts ...ByteStreamWriteOption) (int64, error) {
+	opts := &ByteStreamWriteOpts{FinishWrite: true}
+	for _, o := range rawOpts {
+		o.Apply(opts)
+	}
+
 	var totalBytes int64
 	closure := func() error {
 		// Retry by starting the stream from the beginning.
 		if err := ch.Reset(); err != nil {
 			return errors.Wrap(err, "failed to Reset")
 		}
 		totalBytes = int64(0)
-		// TODO(olaola): implement resumable uploads.
+		// TODO(olaola): implement resumable uploads. ByteStreamWriteOption passed in allows to
+		// upload to an arbitrary offset, but retries still resend from the specified offset.
 
 		stream, err := c.Write(ctx)
 		if err != nil {
 			return err
 		}
 		for ch.HasNext() {
-			req := &bspb.WriteRequest{}
+			req := &bspb.WriteRequest{ResourceName: name}
 			chunk, err := ch.Next()
 			if err != nil {
 				return err
 			}
-			if chunk.Offset == 0 {
-				req.ResourceName = name
-			}
-			req.WriteOffset = chunk.Offset
+			req.WriteOffset = chunk.Offset + opts.LastLogicalOffset
```

Comment:
This would apply to all chunks?

Reply:
Yes but LastLogicalOffset doesn't basically affect WriteOffset in the case where LastLogicalOffset is not passed in as ByteStreamWriteOption. If LastLogicalOffset is not passed in, the value of LastLogicalOffset will be 0 as default and it will be basically req.WriteOffset = chunk.Offset. 

Refactored this code block in this way so that it's easier to read and avoid adding many if statements.

Reply:
Ack